src/models/extra_inputs.cpp

```diff
@@ -1,86 +1,52 @@
-#include "../generators.h"
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+#include "lora_adapter.h"
 #include "model.h"
 #include "extra_inputs.h"
-#include "kernels.h"
 
 namespace Generators {
 
-ExtraInputs::ExtraInputs(const Model& model, State& state)
-    : model_{model},
-      state_{state} {
-  extra_inputs_.reserve(state_.params_->extra_inputs.size());
+ExtraInputs::ExtraInputs(const Model& model, State& state) : model_{model}, state_{state} {
+  auto& lora_management = model_.GetLoraAdapterManagement();
+  const auto total_inputs = state_.params_->extra_inputs.size() + lora_management.GetParamNum();
+  extra_input_names_.reserve(total_inputs);
+  extra_inputs_.reserve(total_inputs);
 
   if (state_.GetCapturedGraphInfo()) {
-    owned_extra_inputs_.reserve(state_.params_->extra_inputs.size());
-
     for (int i = 0; i < state_.params_->extra_inputs.size(); ++i) {
       auto type_and_shape_info = state_.params_->extra_inputs[i].tensor->ort_tensor_->GetTensorTypeAndShapeInfo();
       const auto& input_name = state_.params_->extra_inputs[i].name;
 
-      sb_extra_inputs_.emplace(input_name, state_.GetCapturedGraphInfo()->sb_extra_inputs_.at(input_name).get());
-      owned_extra_inputs_.push_back(sb_extra_inputs_.at(input_name)->CreateTensorOnStaticBuffer(type_and_shape_info->GetShape(), type_and_shape_info->GetElementType()));
-      extra_inputs_.push_back(owned_extra_inputs_.back().get());
+      auto* sb_extra = state_.GetCapturedGraphInfo()->sb_extra_inputs_.at(input_name).get();
+      auto ort_value =
+          sb_extra->CreateTensorOnStaticBuffer(type_and_shape_info->GetShape(), type_and_shape_info->GetElementType());
+
+      // Copy to value created on top of the StaticBuffer
+      CopyToDevice(model_, *state_.params_->extra_inputs[1].tensor->ort_tensor_, *ort_value);
```

Comment:
Why do we need a real Copy here?

Reply:
Copying to the device was a part of the original PR https://github.com/microsoft/onnxruntime-genai/pull/464 which introduced CapturedGraphInfo. You have reviewed and approved it.

It seems that when the model is ran on device, then those extra inputs need to be on device. I am not sure about the exact nature of those, they seem to be special. The memory allocated is actually x times bigger than the shape, apparently related to BeamSearch.